File: 1-EASY/Python/9-word-frequency-counter/proposed-solution/main.py
# Challenge: Implement a function that, given a paragraph or a long string, returns a dictionary with the frequency of each word.
# Handle case differences and punctuation correctly.

import string

def word_frequency_counter(text):
    # Normalize text: lowercase and remove punctuation
    translator = str.maketrans('', '', string.punctuation)
    normalized = text.lower().translate(translator)
    words = normalized.split()
    frequency = {}
    for word in words:
        frequency[word] = frequency.get(word, 0) + 1
    return frequency

# Example usage
if __name__ == "__main__":
    sample_text = "Hello, hello! How are you? Are you doing well? Hello!"
    freq = word_frequency_counter(sample_text)
    for word, count in freq.items():
        print(f"'{word}': {count}")

File: 1-EASY/JavaScript/9-word-frequency-counter/proposed-solution/main.js
// Challenge: Create a function that counts the frequency of each word in a given text.
// Consider normalization of case and removal of punctuation.

function wordFrequency(text) {
    const punctuationRegex = /[!"#$%&'()*+,\-./:;<=>?@[\\\]^_`{|}~]/g;
    text = text.toLowerCase().replace(punctuationRegex, '');
    const words = text.split(/\s+/);
    const freq = {};
    for (const word of words) {
        if (word.length === 0) continue;
        freq[word] = (freq[word] || 0) + 1;
    }
    return freq;
}

// Example usage
const sampleText = "Hello world! Hello everyone. Welcome to the world of Python.";
const frequencies = wordFrequency(sampleText);
for (const [word, count] of Object.entries(frequencies)) {
    console.log(`'${word}': ${count}`);
}
